appmigrator_template.sh

```diff
@@ -0,0 +1,205 @@
+    sleep 2
+    kill $tail_pid || true
+    exit ${exit_code}
+}
+
+#######################################
+# Globals:
+#   CONFIG_PATH
+# Arguments:
+#   var_name: the name of the required entry in config.json
+# Returns:
+#   var_value: the value of the entry
+#######################################
+resin_var_manual_load() {
+    local var_name=$1
+    local var_value
+
+    if [ -f "${CONFIG_PATH}" ]; then
+        var_value=$(jq -r ".${var_name}" "${CONFIG_PATH}") || finish_up "Couldn't get ${var_name} from config.json."
+        if [ -z "${var_value}" ]; then
+            finish_up "Couldn't load ${var_name} key manually."
+        fi
+    else
+        finish_up "Couldn't find the config.json file."
+    fi
+    echo "${var_value}"
+}
+
+main() {
+    local current_release_info
+    local target_release_info
+    local found_from_volume=false
+    local found_to_volume=false
+
+    workdir="/mnt/data/appmigrate"
+    mkdir -p "${workdir}" && cd "${workdir}"
+
+    # also sets tail_pid
+    setup_logfile "${workdir}"
+
+    # Fill in missing global variables, mostly for 2.0.0-rcX OS versions, that have problem with "resin-var"
+    if [ -z "${API_ENDPOINT}" ]; then
+        API_ENDPOINT=$(resin_var_manual_load "apiEndpoint")
+    fi
+    if [ -z "${DEVICE_ID}" ]; then
+        DEVICE_ID=$(resin_var_manual_load "deviceId")
+    fi
+
+    # Application ID check
+    APPLICATION_ID=$(curl --silent --fail --retry 5 -X GET \
+    "${API_ENDPOINT}/v4/device(${DEVICE_ID})" \
+    -H "Content-Type: application/json" \
+    -H "Authorization: Bearer ${TOKEN}" | jq -r '.d[0].belongs_to__application.__id')
+    if [ -z "${APPLICATION_ID}" ]; then
+        finish_up "Couldn't get Application ID"
+    fi
+    if [ "${APPLICATION_ID}" == "${TARGET_APP_ID}" ]; then
+        finish_up "Current application is the same as the target."
+    fi
+
+    current_release_info=$(curl --silent --fail --retry 5 -X GET \
+    "${API_ENDPOINT}/v4/release?\$filter=belongs_to__application%20eq%20${APPLICATION_ID}" \
+    -H "Content-Type: application/json" \
+    -H "Authorization: Bearer $TOKEN") || finish_up "Couldn't get release data."
+    # If the user has no access to the app, the reply will be empty
+    local -a current_volumes=()
+    while IFS='' read -r volume; do
+        current_volumes+=("$volume")
+    done < <(echo "${current_release_info}" | jq -r '.d[0].composition.volumes | keys[]')
+    if [ ${#current_volumes[@]} -eq 0 ]; then
+        finish_up "Could not extract current volumes."
+    fi
+
+    echo "Current volumes:"
+    for v in "${current_volumes[@]}"; do
+        echo -n "$v"
+        if [ "${v}" == "${FROM_VOLUME:=resin-data}" ]; then
+            found_from_volume=true
+            echo " (selected)"
+        else
+            echo
+        fi
+    done
+    if [ "${found_from_volume}" != "true" ]; then
+        finish_up "Couldn't select volume to back up from."
+    fi
+
+    target_release_info=$(curl --silent --fail --retry 5 -X GET \
```

Comment:
I wonder if it's worth generalizing this base curl command since it's fairly standard.

Reply:
yeah, would be nicer that way, will give that a shot.